rados/rados_set_locator.go

```diff
@@ -0,0 +1,27 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+)
+
+// SetLocator sets the key for mapping objects to pgs within an io context.
+// Until a different locator key is set, all objects in this io context will be placed in the same pg.
+// To reset the locator, an empty string must be set.
+//  PREVIEW
+//
+// Implements:
+//  void rados_ioctx_locator_set_key(rados_ioctx_t io, const char *key);
+func (ioctx *IOContext) SetLocator(locator string) {
+	var c_loc *C.char = C.CString(locator)
+	defer C.free(unsafe.Pointer(c_loc))
+	C.rados_ioctx_locator_set_key(ioctx.ioctx, c_loc)
```

Comment:
`rados_ioctx_locator_set_key` expects a `NULL` as `key` if you want to reset the locator, but `C.CString()` does not return `nil` for an empty string. You have to check, if the string is empty, and only if not, you allocate the c-string and assign it to c_loc.